sigmadsp/communication/sigmastudio_protocols.py

```diff
@@ -0,0 +1,398 @@
+"""This module contains implementations for SigmaStudio TCPIP* protocol decoding.
+
+These handle the packet decoding and conditioning, while actual communication happens upstream.
+
+Protocol headers for ADAU144x/5x/6x are documented on the Analog Devices wiki at
+https://wiki.analog.com/resources/tools-software/sigmastudio/usingsigmastudio/tcpipchannels
+
+Protocol headers for ADAU1401/1701 are documented on the Analog Devices forum at
+https://ez.analog.com/dsp/sigmadsp/f/q-a/163849/adau1401-adau1701-tcp-ip-documentation-unonficial-self-made
+"""
+import socket
+from abc import ABC
+from typing import Dict, Tuple, Type, Union
+
+from sigmadsp.helper.conversion import bytes_to_int, int_to_bytes
+
```

Comment:
I have a rather lengthy suggestion, which would make the code easier to follow (in my opinion), and allows for error checks. I would introduce classes for a `Field` and the collection of fields, which is `HeaderFields`.

Let me know what you think:

```suggestion
@dataclass
class Field:
    # The name of the field, e.g. "operation", "safeload", "channel".
    name: str

    # The offset of the field in bytes from the start of the header.
    offset: int

    # The size of the field in bytes.
    size: int

    # The stored value.
    value: int = 0

    def __post_init__(self):
        if self.size < 0:
            raise ValueError("Field size must be a positive integer.")

        if self.offset < 0:
            raise ValueError("Field offset must be a positive integer.")

        self.end = self.offset + self.size - 1


class HeaderFields:
    def __init__(self):
        self._fields: Set[Field] = set()

    @property
    def fields(self) -> Set[Field]:
        """The fields of the header, sorted by their offset."""
        return sorted(self._fields, key=lambda field: field.offset)

    @property
    def length(self) -> int:
        """The total length of the header."""
        return sum([field.size for field in self.fields])

    @property
    def continuous(self) -> bool:
        """Whether or not there are spaces in the header that are not defined."""
        for field, next_field in zip(self.fields, self.fields[1:]):
            if (field.end + 1) != next_field.offset:
                return False

        return True

    def _sanity_check(self):
        """Check for overlapping fields.

        Raises:
            MemoryError: If overlapping fields are found.
        """
        # Check for overlapping fields, which are sorted by their offset.
        for field, next_field in zip(self.fields, self.fields[1:]):
            if field.end <= next_field.offset:
                raise MemoryError("Fields {field.name} and {next_field.name} overlap.")

    def add(self, field: Field):
        """Add a new field. Duplicates are ignored.

        Args:
            field (Field): The field to add.
        """
        self._fields.append(field)

    def get_field_by_name(self, name: str) -> Union[Field, None]:
        """Get a field by its name.

        Args:
            name (str): The name of the field.

        Returns:
            Union[Field, None]: The field, or None, if no field was found.
        """
        for field in self._fields:
            if field.name == name:
                return field

        return None
```

Reply:
it's overly complicated, how do you even use that... anyway, I'll apply it if you give me a hint on how it's used; I just want to get this merged and then never touch this style of Python again...

Reply:
As I say, it is only a suggestion and you don't have to use it. I just find it easier to read than dictionaries, where the fields have no names.

Reply:
Well, you'll be stuck with it long term, if your mind works better with those dataclasses and class hierarchies when dicts are easy to read at these dimensions and give you ordering for free. I find this style exhausting to read (with the configuration parser it took me like half an hour to figure out how I should name my volume control, and I still got it wrong because it seemed to me `target` had to be in the name), and half the time I spent coding was spent figuring out clever ways to avoid import loops and making a clean language look like a very ugly version of Go...

Can you make the suggestion checks-compliant and add an usage example in the `HeaderFields` class docstring? I'm not being passive aggressive, I genuinely have no idea how I'd use this (and what I'd add to the docstrings because at least one of those annoying checks will complain).

Reply:
> with the configuration parser it took me like half an hour to figure out how I should name my volume control, and I still got it wrong because it seemed to me `target` had to be in the name

Well, I am not saying it's perfect. The variable naming can certainly be improved in many areas.


> Can you make the suggestion checks-compliant and add an usage example in the `HeaderFields` class docstring?

I don't want to force you to adopt my `dataclass`-heavy style, so I created a pull request on top of yours. See https://github.com/m0n5t3r/sigmadsp/pull/1. The usage should be very similar to the previous dicts, but adds some comfort methods.

Reply:
merged, confirmed it works; I don't think OrderedDict was strictly necessary (turns out dicts are ordered in Python 3.6 and later), but I guess it helps compatibility with older versions.